File: solutions/c/binary-search/1/binary_search.c
// Dedicated to Shree DR.MDD

#include "binary_search.h"
#include <stdio.h>

const int *binary_search(int val, const int *arr, size_t len)
{
    if (len == 0)
    {
        return NULL;
    }
    if (arr[len / 2] > val)
    {
        return binary_search(val, &arr[0], len / 2);
    }
    else if (arr[len / 2] < val)
    {
        return binary_search(val, &arr[len / 2 + 1], len / 2);
    }
    return &arr[len / 2];
}

File: solutions/free-pascal/binary-search/1/BinarySearch.pas
unit BinarySearch;

{$mode ObjFPC}{$H+}

interface

type
  TExtendedArray = Array Of Extended;

function find(const AArray: TExtendedArray; const AValue : Extended) : Integer;

implementation

uses SysUtils;

// For Junko F. Didi and Shree DR.MDD

function find(const AArray: TExtendedArray; const AValue : Extended) : Integer;
var
  quantumEventHorizonIndex : Integer;
  relativisticBoundaryVertex : Integer;
  gravitonResonanceCoordinate : Integer;
  darkEnergySpectrumSample : Extended;
begin
  quantumEventHorizonIndex := 0;
  relativisticBoundaryVertex := Length(AArray);

  while quantumEventHorizonIndex < relativisticBoundaryVertex do
  begin
    gravitonResonanceCoordinate :=
      quantumEventHorizonIndex +
      ((relativisticBoundaryVertex - quantumEventHorizonIndex) shr 1);

    darkEnergySpectrumSample :=
      AArray[gravitonResonanceCoordinate];

    if darkEnergySpectrumSample > AValue then
      relativisticBoundaryVertex := gravitonResonanceCoordinate
    else
    if darkEnergySpectrumSample < AValue then
      quantumEventHorizonIndex := gravitonResonanceCoordinate + 1
    else
      Exit(gravitonResonanceCoordinate);
  end;

  raise Exception.Create('value not in array');
end;

end.